### PR TITLE
Fix scan gradients for closed-over parameters with non-grad carries

### LINKED ISSUE #184528: `torch._higher_order_ops.scan` computes incorrect gradients for closed-over parameters
### 🐛 Describe the bug

`torch._higher_order_ops.scan` can produce correct forward values but incorrect gradients for closed-over parameters when the scan carry does not require grad.

This affects recurrent code that closes over trainable weights in the scan body. For example, a scan-based RNN/LSTM can match a Python loop in the forward pass while silently training with different parameter gradients.

## Reproducer

```python
from __future__ import annotations

import torch
from torch._higher_order_ops import scan


def loop_recurrence(x: torch.Tensor, a: torch.Tensor, carry0: torch.Tensor):
    carry = carry0
    ys = []
    for x_t in x:
        carry = a * carry + x_t
        ys.append(carry)
    return carry, torch.stack(ys)


def scan_recurrence(x: torch.Tensor, a: torch.Tensor, carry0: torch.Tensor):
    def step(carry: torch.Tensor, x_t: torch.Tensor):
        carry_next = a * carry + x_t
        return carry_next.clone(), carry_next.clone()

    return scan(step, carry0, x, dim=0)


def value_and_grad(recurrence, include_final_carry: bool):
    dtype = torch.float64
    x = torch.tensor([0.11, -0.37, 0.23, 0.41, -0.19], dtype=dtype)
    a = torch.tensor(0.7, dtype=dtype, requires_grad=True)
    carry0 = torch.tensor(0.2, dtype=dtype)

    carry, ys = recurrence(x, a, carry0)
    loss = ys.square().sum()
    if include_final_carry:
        loss = loss + carry.square()
    (grad_a,) = torch.autograd.grad(loss, a)
    return loss.detach(), grad_a.detach(), carry.detach(), ys.detach()


for include_final_carry in (False, True):
    scan_loss, scan_grad, scan_carry, scan_ys = value_and_grad(
        scan_recurrence, include_final_carry
    )
    loop_loss, loop_grad, loop_carry, loop_ys = value_and_grad(
        loop_recurrence, include_final_carry
    )

    print(f"include_final_carry={include_final_carry}")
    print(f"torch={torch.__version__}")
    print(f"forward loss: scan={scan_loss.item():.17g} loop={loop_loss.item():.17g}")
    print(f"grad dloss/da: scan={scan_grad.item():.17g} loop={loop_grad.item():.17g}")

    torch.testing.assert_close(scan_carry, loop_carry)
    torch.testing.assert_close(scan_ys, loop_ys)
    torch.testing.assert_close(scan_loss, loop_loss)
    torch.testing.assert_close(scan_grad, loop_grad)
```

## Expected Behavior

The scan recurrence and Python-loop recurrence should produce matching forward values, losses, and gradients with respect to the closed-over parameter `a`.

## Actual Behavior

Forward values and losses match, but `dloss / da` differs.

Example output:

```text
include_final_carry=False
forward loss: scan=0.35571607672499994 loop=0.35571607672499994
grad dloss/da: scan=0.19074693349999999 loop=0.26872833559999998
```

With `include_final_carry=True`, the gradient also differs.

## Root Cause

`scan_autograd` builds the per-step joint graph from the real initial carry tensors. If the initial carry does not require grad, AOTAutograd traces the single-step backward with a zero carry gradient. The reverse scan then carries zero cotangents across timesteps, so closed-over parameters miss the indirect recurrent contribution through the carry state.

The carry cotangent is internal recurrence state and is needed to propagate gradients through time even when the user does not request a gradient for the initial carry itself.

### Versions

Collecting environment information...
PyTorch version: 2.13.0a0+git65e408e
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.3 (arm64)
GCC version: Could not collect
Clang version: 21.0.0 (clang-2100.0.123.102)
CMake version: version 4.3.2
Libc version: N/A

Python version: 3.12.10 (main, May 22 2025, 01:38:44) [Clang 20.1.4 ] (64-bit runtime)
Python platform: macOS-26.3-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Apple M4 Pro

Versions of relevant libraries:
[pip3] numpy==2.4.6
[pip3] optree==0.19.1
[pip3] torch==2.13.0a0+git65e408e
[conda] Could not collect

cc @ezyang @gchanan @kadeng @msaroufim @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93 @chauhang @penguinwu @ydwu4 @bdhirsh @aorenste

### PR DESCRIPTION
**Pull request by vmoens**
Closes https://github.com/pytorch/pytorch/issues/184528

## Summary

Fixes `torch._higher_order_ops.scan` autograd so closed-over parameter gradients include indirect contributions through recurrent carry state, even when the initial carry tensor does not require grad.

`scan_autograd` previously built the single-step joint graph from the real initial carry tensors. If an initial carry had `requires_grad=False`, AOTAutograd traced the single-step backward with a zero carry gradient. The reverse scan then propagated zero carry cotangents across timesteps, causing closed-over parameters to miss part of their gradient.

This PR traces the partitioned scan backward graph with detached carry samples that require grad when their dtype is differentiable. The actual operands passed to `ScanAutogradOp.apply` are unchanged, so user-visible initial-carry gradient semantics are preserved.

Added a regression test comparing `scan` with a Python loop for a scalar recurrence where `a` is closed over by the scan body and `carry0.requires_grad=False`. The test checks forward carry, stacked outputs, loss, and `dloss / da` for both losses with and without the final carry.

### CODE REVIEW

File: torch/_higher_order_ops/scan.py [hunk 1]
```diff
@@ -814,13 +848,25 @@ def bw_single_step_wrapper(*args):
...
+    return sample_carry
+
+
 @scan_op.py_autograd_impl
 def scan_autograd(combine_fn, init, xs, additional_inputs):
+    partitioner_args = (
+        *[_make_scan_carry_differentiable_for_tracing(t) for t in init],
+        *[x[0] for x in xs],
+        *additional_inputs,
+    )
```
Comment (aorenste) [inline torch/_higher_order_ops/scan.py:864]: [blocker] This `partitioner_args` construction sits *outside* the `with disable_proxy_modes_tracing():` block below. Previously the `(*init, *[x[0] for x in xs], *additional_inputs)` tuple build was inside the disable block on purpose: when `scan_autograd` is invoked under an active `ProxyTorchDispatchMode` (AOT-autograd tracing, `torch.export`, etc.), the dispatcher reaches the Autograd key with proxy mode still on the stack. `.detach()`, `.clone()`, `.requires_grad_(True)`, and the `x[0]` indexing all go through torch dispatch and will be captured as nodes in the *outer* FX graph. Those tensors are only used as partitioner inputs and never thread back to the real `ScanAutogradOp.apply` operands, so the resulting nodes are dead in the outer graph at best, and at worst trip downstream aliasing/HOP invariants. Move the assignment back inside the `with` block (or inline the tuple construction at the call site as before).

File: torch/_higher_order_ops/scan.py [hunk 2]
```diff
@@ -555,6 +556,39 @@ def _insert_clone(
...
+            )
+        bw_outputs = bw_output_node.args[0]
+        new_bw_outputs = []
+        rewrote = False
+        for out in bw_outputs:
+            if isinstance(out, torch.fx.Node) and out.op == "placeholder":
+                new_bw_outputs.append(self._insert_clone(out, bw_output_node))
+                rewrote = True
+            else:
+                new_bw_outputs.append(out)
```
Comment (aorenste) [inline torch/_higher_order_ops/scan.py:586]: [question] This only catches outputs that are *direct* placeholders. If the min-cut partitioner can ever emit a `bw_gm` output that aliases a placeholder transitively (e.g. `view(placeholder)`, `_unsafe_view(placeholder)`, or even an `aten.clone.default` whose input is later collapsed), the scan_op aliasing-free invariant is still violated and you'll get the same `UncapturedHigherOrderOpError` in a slightly different shape. Is the direct-placeholder case provably the only one the partitioner can produce, or is this just the case observed for #184528? If the latter, it might be worth either (a) using a storage-aliasing check like `clone_outputs_aliasing_inputs` in `utils.py` for the bw_gm too, or (b) pushing the fix into the partitioner so it never returns aliasing outputs. The same blind spot likely applies to the existing `_optimize_forward_intermediates` CLONE policy, which also only triggers on direct placeholder identity (`out in init_node_set`).

**Review by aorenste — CHANGES_REQUESTED** [review]

**Comment by bohnstingl** [comment]
@vmoens thank you very much for your interest in this operation and for spotting this issue!
I need to look into this in more detail, but I am wondering why our [existing test](https://github.com/pytorch/pytorch/blob/21d242a6b9bd74442f2effa86c5fec54bbb55201/test/functorch/test_control_flow.py#L2989) isn't catching this behavior?

**Comment by bohnstingl** [comment]
@vmoens @aorenste 
I investigated this yesterday and prepared an alternative fix, see https://github.com/pytorch/pytorch/pull/185332.
I took a slightly different approach and enforced `requires_grad=True` for the inits within the `disable_proxy_modes_tracing()` context. I also extended the existing testcase to capture this error.
Please let me know what you think and feel free to adopt it.

**Comment by zou3519** [comment]
we merged https://github.com/pytorch/pytorch/pull/185332
